linker: make debug monitor isr symbol weak

pjasinski990 opened the pull request:
It seems that currently it's impossible to create a custom implementation for debug monitor exception without updating the vector table (z_arm_debug_monitor maps to fault).
My proposition is to make this symbol weak, so that it can be overriden.

Additionally I added a config for setting the debugmon ISR to low priority (it should be low priority for debugging purposes). 

Signed-off-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

pjasinski990 committed:
linker: make debug monitor isr symbol weak

It seems that currently it's impossible to create a custom
implementation for debug monitor exception without updating the vector
table (z_arm_debug_monitor maps to fault).
My proposition is to make this symbol weak, so that it can be overriden.

Signed-off-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

pjasinski990 committed:
Kconfig: add config for low-priority debug mon isr

Debug monitor needs to be configured to a low priority in order to be
useful for debugging (to prioritize other interrupts when waiting on a
breakpoint).
Added a config that configures the interrupt this way.

Signed-off-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

microbuilder commented:
CI failing consistently, but doesn't seem related:

```
INFO    - 438/477 qemu_cortex_m0            zephyr/samples/modules/chre/sample.modules.chre    FAILED Timeout (qemu 59.955s)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_cortex_m0/zephyr/samples/modules/chre/sample.modules.chre/handler.log
ERROR   - *** Booting Zephyr OS build zephyr-v3.2.0-1908-g7088341374f3 ***
Failed to set thread name to "CHRE": -88
Hello CHRE!
[00:00:00.002,023] <inf> chre: CHRE init, version: 
[00:00:00.002,500] <inf> chre: EventLoop start
EchoApp::nanoappStart()
Starting EchoApp... SUCCESS
Nanoapp count=1
Finding instance ID... SUCCESS
Nanoapp count=1
Instance ID: 1
Sending event 1...
EchoApp::nanoappHandleEvent(sender_instance_id=0, event_type=1, event_data@(nil))
Event (1) complete!
Ending EchoApp... FAIL
Goodbye!
EchoApp::nanoappEnd()
[00:00:01.020,148] <err> chre: Refusing to unload system nanoapp
```

microbuilder commented:
@pjasinski990 It would be nice to extend this with an example, and showing how to configure registers to properly enable the debug monitor (clear `C_DEBUGEN` and set `DBGKEY` in `DHCSR`, etc.)

pjasinski990 committed:
samples: add debug monitor sample

Implemented a sample that shows configuration and basic usage of monitor
mode debugging feature.

Signed-off-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

microbuilder commented:
Given that the entire implementation here https://github.com/nrfconnect/sdk-nrf/pull/9505 is basically just Segger code, wouldn't adding this to Zephyr be better from a maintenance point of view? Given that the code isn't coming from Nordic, it will just end up with a situation where someone will reimplement the same again upstream anyway?

@carlescufi Any thoughts?

carlescufi commented:
> Given that the entire implementation here [nrfconnect/sdk-nrf#9505](https://github.com/nrfconnect/sdk-nrf/pull/9505) is basically just Segger code, wouldn't adding this to Zephyr be better from a maintenance point of view? Given that the code isn't coming from Nordic, it will just end up with a situation where someone will reimplement the same again upstream anyway?
> 
> @carlescufi Any thoughts?

Yes, I agree. Then this Segger code would have to go in the [segger](https://github.com/zephyrproject-rtos/segger) module repo. @pjasinski990 could you make this change please?

pjasinski990 committed:
modules: build segger debugmon code with config

Updated segger module CMake to build code added by
https://github.com/zephyrproject-rtos/segger/pull/14 when relevant
config is provided.

Signed-off-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

pjasinski990 committed:
doc: add monitor mode debugging entry

Added documentation on monitor mode debugging usage.

Signed-off-by: Piotr Jasiński <piotr.jasinski@nordicsemi.no>

carlescufi commented:
> Given that the entire implementation here [nrfconnect/sdk-nrf#9505](https://github.com/nrfconnect/sdk-nrf/pull/9505) is basically just Segger code, wouldn't adding this to Zephyr be better from a maintenance point of view? Given that the code isn't coming from Nordic, it will just end up with a situation where someone will reimplement the same again upstream anyway?
> 
> @carlescufi Any thoughts?

@microbuilder this is now done, thanks for the suggestion, it really makes sense.

microbuilder commented on samples/subsys/debug/debugmon/src/main.c:
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
```
Should this be updated to 2022 and Nordic, since it looks like a new file?

microbuilder commented on samples/subsys/debug/debugmon/src/main.c:
```diff
@@ -0,0 +1,79 @@
+
+int debug_mon_enable(void)
+{
+	/*
+		Cannot enable monitor mode if C_DEBUGEN bit is set. This bit can only be
+		altered from debug access port. It is cleared on power-on-reset.
+	*/
+	bool is_in_halting_mode = (CoreDebug->DHCSR & CoreDebug_DHCSR_C_DEBUGEN_Msk) == 1;
+	if (is_in_halting_mode) {
+		return -1;
+	}
+
+	/* Enable monitor mode debugging by setting MON_EN bit of DEMCR */
+	CoreDebug->DEMCR |= CoreDebug_DEMCR_MON_EN_Msk;
+	return 0;
+}
+
+/*
+	Interrupt handler that will be called each time the device
+	enters a breakpoint while debugging in monitor mode.
+	With CONFIG_DEBUG_MONITOR enabled this exception is set to lowest
+	possible priority (IRQ_PRIO_LOWEST).
+*/
```
These comments will need cleanup to respect the Zephyr style guide

microbuilder commented on subsys/debug/Kconfig:
```diff
@@ -421,3 +421,12 @@ config GDBSTUB_BUF_SZ
 endif
+
+config USE_SEGGER_DEBUGMON
+	bool "Use debug monitor implementation provided by segger for J-LINK debug probes"
```
```suggestion
	bool "Use Segger's J-Link debug monitor implementation"
```

microbuilder commented on subsys/debug/Kconfig:
```diff
@@ -421,3 +421,12 @@ config GDBSTUB_BUF_SZ
+	default false
+	depends on DEBUG_MONITOR
+	help
+	  This option will enable building of SEGGER-provided
+	  implementaiton of debug monitor interrupt and override
+	  z_arm_debug_monitor symbol.
```
```suggestion
	  This option will enable Segger's implementation of
	  the debug monitor interrupt, overriding the
	  default z_arm_debug_monitor symbol.
```

microbuilder commented:
@pjasinski990 Can you clean up the compliance issues? Everything else looks good.

pjasinski990 commented on subsys/debug/Kconfig:
```diff
@@ -421,3 +421,12 @@ config GDBSTUB_BUF_SZ
 
 endif
+
+config USE_SEGGER_DEBUGMON
+	bool "Use Segger's J-Link debug monitor implementation"
+	default false
+	depends on DEBUG_MONITOR
+	help
+        This option will enable Segger's implementation of
+	    the debug monitor interrupt, overriding the
+	    default z_arm_debug_monitor symbol.
```
The implementation is specifically for J-Links, we can leave it as is.

carlescufi commented:
@microbuilder can you please take another look?

carlescufi commented:
@carlocaione @nordic-krch would you mind reviewing?

microbuilder commented:
@carlescufi compliance still failing here

stephanosio commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -96,6 +96,13 @@ config DCACHE_LINE_SIZE
 config ICACHE_LINE_SIZE
 	default 32
 
+config DEBUG_MONITOR
```
DebugMonitor exception is a Cortex-M-specific feature and should be prefixed accordingly.


```suggestion
config CORTEX_M_DEBUG_MONITOR
```

stephanosio commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -96,6 +96,13 @@ config DCACHE_LINE_SIZE
 
+config DEBUG_MONITOR
+	bool "Use debug monitor interrupt for debugging"
+	depends on !CPU_CORTEX_M0
```
This is too specific. This needs to be:

```suggestion
	depends on !ARMV6_M_ARMV8_M_BASELINE
```

stephanosio commented on subsys/debug/Kconfig:
```diff
@@ -421,3 +421,12 @@ config GDBSTUB_BUF_SZ
 
 endif
+
+config USE_SEGGER_DEBUGMON
```
`USE_` is redundant.

```suggestion
config SEGGER_DEBUGMON
```

stephanosio commented on subsys/debug/Kconfig:
```diff
@@ -421,3 +421,12 @@ config GDBSTUB_BUF_SZ
+
+config USE_SEGGER_DEBUGMON
+	bool "Use Segger's J-Link debug monitor implementation"
+	default n
```
`default n` is redundant.

```suggestion
```

microbuilder commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -96,6 +96,13 @@ config DCACHE_LINE_SIZE
+	bool "Use debug monitor interrupt for debugging"
+	depends on !ARMV6_M_ARMV8_M_BASELINE
+	help
+	  Enable this option to configure debug monitor exception to low priority
+	  for debugging purposes.
```
@stephanosio Think the proposed name makes sense, yes.

stephanosio commented on modules/segger/CMakeLists.txt:
```diff
@@ -1,15 +1,29 @@
 if(CONFIG_USE_SEGGER_RTT)
+    zephyr_library()
```
Use two spaces for indentation in CMake files.

stephanosio commented on samples/subsys/debug/debugmon/src/main.c:
```diff
@@ -0,0 +1,84 @@
+ */
+void z_arm_debug_monitor(void)
+{
+    /*
+     * Implement the logic or use a ready implementation available for your platform
+     * (ie. SEGGER implementation for their debug probes)
+     */
```
Need to use TAB

stephanosio reviewed (dismissed):
Looks ok overall. Some leftover style issues need to be fixed.

microbuilder commented:
Compliance looks good, thanks for cleaning that up. Not sure why the BLE tests are failing. @carlescufi ???

carlescufi commented:
@microbuilder @stephanosio can you please take another look?

carlocaione commented on samples/subsys/debug/debugmon/src/main.c:
```diff
@@ -0,0 +1,84 @@
+{
+	/* Set up led and led timer */
+	if (!device_is_ready(led.port)) {
+		printk("%s\n", "Device not ready");
```
this looks odd to me. Why not simply `printk("....");` ?

microbuilder replied:
> Compliance looks good, thanks for cleaning that up. Not sure why the BLE tests are failing. @carlescufi ???

Looks like kicking CI again fixed the failure. Other than the comment from Carlos, I'm OK with this PR, but I'll wait for those changes to approve.

carlescufi commented:
@stephanosio @microbuilder please take another look
